portable/ARMv8M/non_secure/port.c

```diff
@@ -1069,6 +1091,113 @@ void vPortSVCHandler_C( uint32_t * pulCallerStackAddress ) /* PRIVILEGED_FUNCTIO
 
 BaseType_t xPortStartScheduler( void ) /* PRIVILEGED_FUNCTION */
 {
+    #if ( configASSERT_DEFINED == 1 && portARM_CORTEX_M23 != 1 )
+    {
+        volatile uint8_t ucOriginalPriority;
+        volatile uint32_t ulImplementedPrioBits = 0;
+        volatile uint8_t * const pucFirstUserPriorityRegister = ( volatile uint8_t * const ) ( portNVIC_IP_REGISTERS_OFFSET_16 + portFIRST_USER_INTERRUPT_NUMBER );
```

Comment:
We constrain register access to be aligned in most cases. It's possible that a user has CCR_UNALIGN_TRP enabled, which could be triggered by the use of uint8_t here.

Reply:
The uint8_t here looks right to me.  The NVIC_IPRn registers are byte accessible by design (on all ARMv8-M except CM23).

There is a possible issue here when running FreeRTOS on the non-secure side because NVIC_IPRn is RAZ/WI (reads as zero, writes ignored) from non-secure state when the corresponding interrupt targets the secure state.

Reply:
@urutva The fix I would recommend here is to use the SVC priority register instead of the priority register for the first user interrupt.  In ARMv8-M, it looks like the system handler priority registers are required to reflect only the number of implemented PRIO bits.  So the strategy should be the same as you have now.  And there are no issues related to secure/non-secure sides.

![image](https://github.com/FreeRTOS/FreeRTOS-Kernel/assets/61161750/5eeb2821-0e43-4b43-941c-e4b8aef16051)

Plus, the ARMv8-M ports require the SVC interrupt priority to be zero, so you can destroy the contents during your PRIO bit counting, and then the subsequent (existing) code that sets for SysTick and PendSV can also include setting SHPR2 to zero.  Be careful though because CM23 doesn't require the PRI_11 field to be byte accessible.  Luckily PRI_10 PRI_9 and PRI_8 are all RES0, so you can safely write zeros with no side effects.

Reply:
When we were testing this, we realized this issue that using NVIC_IPR registers does not work if the secure software does not mark the interrupt as non-secure using the corresponding NVIC_ITNS register. This suggestion looks really good. Let me try that.

Reply:
I tested and it works perfectly. Thank you for the suggestion @jefftenney !

Reply:
👍🏼  One minor nit though, and it could wait for future work.  Now using SHPR2, we shouldn't preserve its value.  The ARMv8-M ports expect SVCall to use priority zero (the highest priority).  The existing code that sets the PendSV and SysTick interrupt priorities should also set the SVCall priority to zero so all the kernel interrupt priorities are set there.

Reply:
@jefftenney Can you please elaborate bit more on the requirement to set SVCall to priority zero (highest priority) for Armv8-M port? Thanks

Reply:
@urutva Here's a [recent forum thread](https://forums.freertos.org/t/xportstartscheduler-fails-because-basepri-is-non-zero/15509) for ARMv7-M.  I think all the same reasoning applies to ARMv8-M.

The SVC instruction induces an exception that must be handled immediately.  Compare to PendSV which is willing to pend for later execution.  If the SVC instruction executes while the current execution priority prevents the SVCall handler, then a fault occurs (usage fault I think).  The forum post describes how the ARMv7-M port executes the SVC instruction with BASEPRI masking some interrupts.  The same happens in ARMv8-M.  I am happy to find some references in ARM docs if it helps.